### PR TITLE
feat: torch open model added

### LINKED ISSUE #38: Попробовать использовать открытую модель для gaze vector estimation
Основная мысль при рассмотрении открытой модели - возможность подтягивать ее веса и дообучать модель под новые данные

вот [тут](https://github.com/yakhyo/gaze-estimation), кажется, есть нужная модель

### LINKED ISSUE #100: Поменять расположение моделей в папке
Супер кринж /intel/intel, надо сделать /models/intel и /models/other имхо

### PR DESCRIPTION
**Pull request by Berletm**

### CODE REVIEW

File: tracking/src/gaze_estimator.py [hunk 1]
```diff
@@ -177,8 +196,54 @@ def __estimate_gaze_vec(self, eyes: Tuple, angles: np.ndarray) -> np.ndarray:
...
+
+        yaw_probs   = torch.softmax(yaw_logits,   dim=1)
+        pitch_probs = torch.softmax(pitch_logits, dim=1)
+        
+        idx_tensor = torch.arange(90, dtype=torch.float32, device=device)
+        
+        yaw_pred   = torch.sum(yaw_probs   * idx_tensor, dim=1) 
+        pitch_pred = torch.sum(pitch_probs * idx_tensor, dim=1)
+        
+        yaw_deg   = yaw_pred   * 4.0 - 180.0
```
Comment (AlexanderKamynin) [inline tracking/src/gaze_estimator.py:235]: Нужно добавить комментарии, поясняющие формулы

File: tracking/src/constants.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-PTH2MODELS = "/intel/intel"
+PTH2MODELS = "/intel/"
```
Comment (AlexanderKamynin) [inline tracking/src/constants.py]: нужно будет поправить пути на актуальные
@Berletm

File: tracking/src/gaze_estimator.py [hunk 3]
```diff
@@ -177,8 +196,54 @@ def __estimate_gaze_vec(self, eyes: Tuple, angles: np.ndarray) -> np.ndarray:
         l = np.linalg.norm(output)
         output /= l if l != 0 else 1
         
+        output *= np.array([1, -1, 1])
+        
         return output
 
+    @torch.no_grad
```
Comment (AlexanderKamynin) [inline tracking/src/gaze_estimator.py]: Декоратор лучше со скобками  @torch.no_grad()

File: tracking/src/gaze_estimator.py [hunk 4]
```diff
@@ -177,8 +196,54 @@ def __estimate_gaze_vec(self, eyes: Tuple, angles: np.ndarray) -> np.ndarray:
...
+    def __estimate_gaze_vec_torch(self, face: np.ndarray) -> np.ndarray:
+        if any([s == 0 for s in face.shape]):
+            return [0, 0, 0]
```
Comment (AlexanderKamynin) [inline tracking/src/gaze_estimator.py]: обещаем вернуть массив, возвращаем список - поправить

File: tracking/src/gaze_estimator.py [hunk 5]
```diff
@@ -177,8 +196,54 @@ def __estimate_gaze_vec(self, eyes: Tuple, angles: np.ndarray) -> np.ndarray:
...
         return output
 
+    @torch.no_grad
+    def __estimate_gaze_vec_torch(self, face: np.ndarray) -> np.ndarray:
+        if any([s == 0 for s in face.shape]):
+            return [0, 0, 0] 
+        
+        transform = T.Compose([
+        T.ToPILImage(),                     
+        T.Resize(448),
```
Comment (AlexanderKamynin) [inline tracking/src/gaze_estimator.py]: Нужно обосновать размер изображения - почему например не (448,448)?

File: tracking/src/gaze_estimator.py [hunk 6]
```diff
@@ -177,8 +196,54 @@ def __estimate_gaze_vec(self, eyes: Tuple, angles: np.ndarray) -> np.ndarray:
...
+        
+        yaw_rad = np.deg2rad(yaw_deg.item())
+        pitch_rad = np.deg2rad(pitch_deg.item())
+        
+        x = -np.cos(pitch_rad) * np.sin(yaw_rad)
+        y = -np.sin(pitch_rad)
+        z = np.cos(pitch_rad) * np.cos(yaw_rad)
+        
+        gaze_vec = np.array([x, y, z])
+        len = np.linalg.norm(gaze_vec)
```
Comment (AlexanderKamynin) [inline tracking/src/gaze_estimator.py]: лучше len заменить на norm

File: tracking/src/gaze_estimator.py [hunk 7]
```diff
@@ -177,8 +196,54 @@ def __estimate_gaze_vec(self, eyes: Tuple, angles: np.ndarray) -> np.ndarray:
...
+        output *= np.array([1, -1, 1])
+        
         return output
 
+    @torch.no_grad
+    def __estimate_gaze_vec_torch(self, face: np.ndarray) -> np.ndarray:
+        if any([s == 0 for s in face.shape]):
+            return [0, 0, 0] 
+        
+        transform = T.Compose([
```
Comment (AlexanderKamynin) [inline tracking/src/gaze_estimator.py]: есть смысл этот трансформ вынести в __init__, если он всегда используется один и тот же

File: tracking/src/gaze_mapper.py [hunk 8]
```diff
@@ -55,7 +55,8 @@ def calibrate(n_epochs: int, model: GazeMapper, train_loader: DataLoader, val_lo
     model = model.to(device)
     
     criterion = nn.MSELoss()
-    optimizer = torch.optim.SGD(model.parameters(), lr=1e-3)
+    optimizer = torch.optim.Adam(model.parameters(), lr=0.3)
```
Comment (AlexanderKamynin) [inline tracking/src/gaze_mapper.py:58]: Для адама обычно используются более маленькие lr. В текущем варианте вероятнее всего обучение будет нестабильным. Попробуй lr=1e-3, максимум 1e-2 - больше перебирать нет смысла

File: tracking/src/gaze_mapper.py [hunk 9]
```diff
@@ -55,7 +55,8 @@ def calibrate(n_epochs: int, model: GazeMapper, train_loader: DataLoader, val_lo
...
+    scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, factor=0.5)
```
Comment (AlexanderKamynin) [inline tracking/src/gaze_mapper.py:59]: Планировщик не используется нигде

Comment (AlexanderKamynin) [reply tracking/src/gaze_mapper.py:59]: Я бы пробовал варианты CosineAnnealingLR или Linear с warmup-ом - но это скорее как идея, чем как замечание

File: tracking/src/gaze_mapper.py
Comment (AlexanderKamynin) [inline tracking/src/gaze_mapper.py:115]: Переменные буковками не обзывают(

File: tracking/src/gaze_mapper.py [hunk 10]
```diff
@@ -117,7 +118,7 @@ def calibrate(n_epochs: int, model: GazeMapper, train_loader: DataLoader, val_lo
     train_loader = DataLoader(train, batch_size=1, shuffle=True, num_workers=2, pin_memory=True)
     val_loader = DataLoader(val, batch_size=1, shuffle=True, num_workers=2, pin_memory=True)
     
-    epochs = 25
+    epochs = 200
```
Comment (AlexanderKamynin) [inline tracking/src/gaze_mapper.py:121]: Сколько по времени калибруется один видос на 200 эпох? Также нужно зафиксировать размер изображения (основная мысль: если мы ждем "качественную" калибровку дольше обработки самого видео (минут 10) - это довольно долго)

File: tracking/src/gaze_mapper.py [hunk 11]
```diff
@@ -117,7 +118,7 @@ def calibrate(n_epochs: int, model: GazeMapper, train_loader: DataLoader, val_lo
     train_loader = DataLoader(train, batch_size=1, shuffle=True, num_workers=2, pin_memory=True)
```
Comment (AlexanderKamynin) [inline tracking/src/gaze_mapper.py:118]: Для воспроизводимости результатов нужно зафиксировать seed

**Review by AlexanderKamynin — CHANGES_REQUESTED** [review]

File: tracking/src/gaze_mapper.py (hunk 10 above)
Comment (AlexanderKamynin) [reply tracking/src/gaze_mapper.py:121]: Время калибровки нужно добавить в описание к задаче

**Comment by AlexanderKamynin** [comment]
Нужно будет также подтянуть фикс с необучаемой матрицей здесь

**Review by AlexanderKamynin — APPROVED** [review]
